### ISSUE TITLE
stdin redirection hangs on macOS

### ISSUE THREAD

**Issue by alexkornitzer** [oven-sh/bun#10080]
Since Bun v1.0.31 which had the major IO rework, talking to Bun via an Erlang Port (https://www.erlang.org/doc/tutorial/c_port) has resulted in hangs on macOS. After investigation the issue appears to come from assumptions that Bun makes in its `FileReader` (https://github.com/oven-sh/bun/blob/main/src/bun.js/webcore/streams.zig#L3441), resulting in it setting the reader as non blocking when the file descriptor is not. This then results in a hang as the `PosixBufferedReader` handles reads in a non blocking way while the read syscall is blocking, thus `EAGAIN` is never raised.

I have used a rather forceful approach in local testing to fix the issue which just ensures the file descriptor is non blocking (using the already present but what appears to be unused function).

```patch
diff --git a/src/bun.js/webcore/streams.zig b/src/bun.js/webcore/streams.zig
index 7703efa21..201d21a83 100644
--- a/src/bun.js/webcore/streams.zig
+++ b/src/bun.js/webcore/streams.zig
@@ -3525,6 +3525,11 @@ pub const FileReader = struct {
                 if (this.nonblocking and this.file_type == .pipe) {
                     this.file_type = .nonblocking_pipe;
                 }
+
+                // ensure non blocking otherwise hangs will occur
+                if (this.nonblocking) {
+                    bun.ensureNonBlocking(fd.int());
+                }
             }

             this.fd = fd;
```

I have not raised a PR for this as I understand the fix might be to handle the broken assumptions for macOS builds rather than forcing the descriptor to be non blocking.

The following was used to reproduce the error:

#### test.js
```javascript
for await (const line of console) {
  console.log(`line: ${line}`);
}
```

#### wrapper.sh
```bash
#!/usr/bin/env bash

exec &> >(./build/bun-debug ./test.js)

echo "I printed...!"

while read -r line; do echo $line; done
```

**Comment by postnerd** [comment]
I can also reproduce this behavior using Bun as the runtime for my chess engine, [postbot](https://github.com/postnerd/postbot/tree/bunjs).

Since Bun v1.0.31, the engine has stopped working with BanksiaGUI or the [lichess-bot wrapper](https://github.com/lichess-bot-devs/lichess-bot). According to the logs, it appears that stdin stops working after the first input.

**Comment by HuakunShen** [comment]
I have a similar issue with Tauri.
I built a RPC protocol to communicate between Tauri's shell API and JS runtime (Nodejs, Deno and Bun).
Bun on Mac is the only scenario that doesn't work. (it works on windows and linux).

Here is the IO adapter implementation for Bun https://github.com/kunkunsh/kkrpc/blob/df3f86d524903999e90de03751716f41cc34dfdd/packages/kkrpc/src/adapters/bun.ts#L23

Technically the Nodejs IO Adapter should also work for Bun
https://github.com/kunkunsh/kkrpc/blob/df3f86d524903999e90de03751716f41cc34dfdd/packages/kkrpc/src/adapters/node.ts#L13

Both adapters fail to read stdin. If I run the same code with deno, bun or node regularly (i.e. not from Tauri app), it works fine.

Here is how bun is spawned in Tauri shell.
https://github.com/kunkunsh/kkrpc/blob/df3f86d524903999e90de03751716f41cc34dfdd/examples/tauri-demo/src/routes/%2Bpage.svelte#L29

I've run the Tauri sample project on Windows, Linux and Mac, Mac is the only one that fails to read anything from stdin.

**Comment by chirayuk** [comment]
Bun v1.2.5.

I believe that I'm experiencing this for a "git-remote-customprotocol" helper.  I tried it as both a bun invocation and as a compiled binary.  Piping to it via `cat` works.  When `git` pipes to it, it hangs on reading from stdin.  The read is done using `for await (const line of createInterface({stdin: process.stdin, ...}) ...`.

**Comment by chirayuk** [comment]
May be related to #18239

**Comment by chirayuk** [comment]
Yes, this is fixed and works great in the latest version.  Thanks!
